force-app/main/default/classes/OpportunityHelper.cls

```diff
@@ -0,0 +1,182 @@
+        Map<Id, Contact> accountIdToContact = new Map<Id, Contact>();
+
+        for (Contact cont : contacts.values()) {
+            if (!accountIdToContact.containsKey(cont.AccountId)) {
+                accountIdToContact.put(cont.AccountId, cont);
+            }
+        }
+
+        for(Opportunity opp : opps){
+            if (accountIdToContact.containsKey(opp.AccountId)) {
+                Contact ceoContact = accountIdToContact.get(opp.AccountId);
+                opp.Primary_Contact__c = ceoContact.Id; 
+            }
+        }
+    }
+
+    //Append stage changes to Opp description after update
+    public static void updateOpportunityDescription(List<Opportunity> opps, Map<Id, Opportunity> oldMap)  {
+        List<Opportunity> oppsToUpdate = new List<Opportunity>();
+        for (Opportunity opp : opps){
+            Opportunity oldOpp = oldMap.get(opp.Id);
+            if (opp.StageName != oldOpp.StageName){
+                oppsToUpdate.add( new Opportunity(
+                    Id = opp.Id,
+                    Name = opp.Name,
+                    StageName= opp.StageName,
+                    Description = 'Stage Change:' + opp.StageName + ':' + DateTime.now().format()
+                ));
+            }
+        }
+        update as user oppsToUpdate;
+        
+    }
+
+    // send notification when opportunity is deleted
+    public static void sendNotificationOpportunityDeleted(List<Opportunity> deletedOpps) {
+        notifyOwnersOpportunityDeleted(deletedOpps);
+    }
+    // Assign the primary contact to undeleted Opportunities after undelete
+    public static void assignPrimaryContactAfterUndeleted(Map<Id,Opportunity> opps) {
+        assignPrimaryContact(opps, 'VP Sales');
+    }
+
+    // method to send the notification
+    public static void notifyOwnersOpportunityDeleted(List<Opportunity> opps) {
+        //get id's from Opp owners
+        Set<Id> userIds = new Set<Id>();
+        for (Opportunity opp : opps) {
+            userIds.add(opp.OwnerId);
+        }
+        // get emails
+        List<User> queriedUsers = [SELECT Id, Email FROM User WHERE Id IN :userIds WITH SECURITY_ENFORCED];
+        
+        Map<Id, String> userEmails = new Map<Id, String>();
+        for (User usr : queriedUsers) {
+            userEmails.put(usr.Id, usr.Email);
+        }
+        List<Messaging.SingleEmailMessage> mails = new List<Messaging.SingleEmailMessage>();
+        for (Opportunity opp : opps){
+            Messaging.SingleEmailMessage mail = new Messaging.SingleEmailMessage();
+            List<String> toAddress = new List<String>();
+            String ownerEmail = userEmails.get(opp.OwnerId);
+            toAddress.add(ownerEmail);
+            mail.setToAddresses(toAddress);
+            mail.setSubject('Opportunity Deleted : ' + opp.Name);
+            mail.setPlainTextBody('Your Opportunity: ' + opp.Name +' has been deleted.');
```

Comment:
Try moving this into a separate method so that is easier to maintain and understand for the future.